llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -14064,11 +14070,24 @@ Value *BoUpSLP::vectorizeTree(
       }
 #endif
       LLVM_DEBUG(dbgs() << "SLP: \tErasing scalar:" << *Scalar << ".\n");
-      eraseInstruction(cast<Instruction>(Scalar));
+      auto *I = cast<Instruction>(Scalar);
+      // Clear the operands, marking for deletion trivially dead operands.
+      for (unsigned Idx : seq<unsigned>(I->getNumOperands())) {
+        // Ignore pointer operand of stores to keep correct DIAssignID.
+        if (isa<StoreInst>(I) && Idx == 1)
+          continue;
+        Value *Op = I->getOperand(Idx);
+        I->setOperand(Idx, PoisonValue::get(Op->getType()));
+        if (auto *OpI = dyn_cast<Instruction>(Op))
+          if (!isDeleted(OpI) && isInstructionTriviallyDead(OpI, TLI) &&
+              Entry->VectorizedValue != OpI)
+            eraseInstruction(OpI);
```

Comment:
Similar code to the version in HorizontalReduction - pull out into a small helper?